src/libraries/Microsoft.PowerFx.Interpreter/Marshal/TypeCoercionProvider.cs

```diff
@@ -32,16 +33,73 @@ internal static FormattingInfo CreateFormattingInfo(RuntimeConfig runtimeConfig,
         };
 
         /// <summary>
-        /// Try to convert value to Boolean format.
+        /// Can convert value to source format to target or not.
+        /// </summary>
+        /// <param name="source">Source type format.</param>
+        /// <param name="target">Target type format.</param>
+        /// <returns>True/False based on whether function can convert from source type to target type.</returns> 
+        public static bool CanPotentiallyCoerceTo(this FormulaType source, FormulaType target)
+        {
+            if (source == FormulaType.Boolean)
+            {
+                return BooleanValue.AllowedListConvertToBoolean.Contains(target);
+            }
+            else if (source == FormulaType.String)
+            {
+                return StringValue.AllowedListConvertToString.Contains(target);
+            }
+            else if (source == FormulaType.Number)
+            {
+                return NumberValue.AllowedListConvertToNumber.Contains(target);
+            }
+            else if (source == FormulaType.DateTime)
+            {
+                return DateTimeValue.AllowedListConvertToDateTime.Contains(target);
+            }
+
+            return false;
+        }
+
+        /// <summary>
+        /// Try to convert value to target format.
         /// </summary>
         /// <param name="value">Input value.</param>
+        /// <param name="targetType">Target type format.</param>
         /// <param name="result">Result value.</param>
-        /// <returns>True/False based on whether function can convert from original type to Boolean type.</returns> 
-        public static bool TryCoerceTo(this FormulaValue value, out BooleanValue result)
+        /// <returns>True/False based on whether function can convert from original type to target type.</returns> 
+        public static bool TryCoerceTo(this FormulaValue value, FormulaType targetType, out FormulaValue result)
         {
-            return TryGetBoolean(IRContext.NotInSource(FormulaType.Boolean), value, out result);
-        }
+            result = null;            
+            if (!value.Type.CanPotentiallyCoerceTo(targetType))
+            {
+                return false;
+            }
```

Comment:
curious - is this necessary? 
Or will the subsequent TryCoerceTo on 81,86,91 just fail?

Reply:
or we could call this at end just to confirm the results are the same; and throw an exception (our bug) if they're different.

Reply:
not necessary. just think we can do check before do any. but i can remove it.